### ISSUE TITLE
None is allowed in generic instantiation regardless of bound

### ISSUE THREAD

**Issue by bluetech** [python/mypy#6039]
This is a little (inconsequential) oddity I've noticed. It seems like `None` bypasses any generic bounds even in strict mode.

* Are you reporting a bug, or opening a feature request?

Bug.

* Please insert below the code you are checking with mypy.

```python
from typing import TypeVar, Generic

T = TypeVar('T', bound=str)  # str or any other type except None
class G(Generic[T]):
    pass

# error: Type argument "builtins.int" of "C" must be a subtype of "builtins.tuple[Any]"
class CheckItNotNone(G[int]):
    pass

# No error.
class CheckItNone(G[None]):
    pass
```

* What is the actual behavior/output?

Annotated in the snippet.

* What is the behavior/output you expect?

`CheckItNone` should emit an error `Type argument "builtins.NoneType" of "C" must be a subtype of "builtins.str"`, similar to the not-None case.

* What are the versions of mypy and Python you are using?

mypy ~~0.641~~ 0.650, Python 3.5.6

* What are the mypy flags you are using? (For example --strict-optional)

`--strict --strict-optional`

**Comment by ilevkivskyi** [comment]
I think this is related to how strict optional is implemented. Although it is technically a default now, `experiments.STRICT_OPTIONAL` is still `False`. Then the per-file strict optional is implemented by setting this "constant" in a context manger _only_ when visiting a file by type checker. During semantic analysis it is always `False`, and this particular check happens at the very end of semantic analysis.

See also https://github.com/python/mypy/issues/5603

**Comment by ilevkivskyi** [comment]
Note that this is fixed in the new semantic analyzer.

**Comment by ilevkivskyi** [comment]
Closing as this is fixed in new semantic analyzer (to be released soon).
